packages/lodestar/src/network/peers/discover.ts

```diff
@@ -0,0 +1,163 @@
+import PeerId from "peer-id";
+import {IBeaconConfig} from "@chainsafe/lodestar-config";
+import {ILogger} from "@chainsafe/lodestar-utils";
+import {Discv5, Discv5Discovery} from "@chainsafe/discv5";
+import {shuffle} from "../../util/shuffle";
+import {getConnectedPeerIds} from "./utils";
+import {IPeerRpcScoreStore, ScoreState} from "./score";
+
+export type AttSubnetQuery = {subnetId: number; maxPeersToDiscover: number};
+
+export type PeerDiscoveryOpts = {
+  maxPeers: number;
+};
+
+export type PeerDiscoveryModules = {
+  libp2p: LibP2p;
+  peerRpcScores: IPeerRpcScoreStore;
+  logger: ILogger;
+  config: IBeaconConfig;
+};
+
+/**
+ * PeerDiscovery discovers and dials new peers, and executes discv5 queries.
+ * Currently relies on discv5 automatic periodic queries
+ */
+export class PeerDiscovery {
+  private libp2p: LibP2p;
+  private peerRpcScores: IPeerRpcScoreStore;
+  private logger: ILogger;
+  private config: IBeaconConfig;
+
+  /** The maximum number of peers we allow (exceptions for subnet peers) */
+  private maxPeers: number;
+
+  constructor(modules: PeerDiscoveryModules, opts: PeerDiscoveryOpts) {
+    this.libp2p = modules.libp2p;
+    this.peerRpcScores = modules.peerRpcScores;
+    this.logger = modules.logger;
+    this.config = modules.config;
+    this.maxPeers = opts.maxPeers;
+  }
+
+  /**
+   * Request to find peers. First, looked at cached peers in peerStore
+   */
+  discoverPeers(maxPeersToDiscover: number): void {
+    const notConnectedPeers = Array.from(this.libp2p.peerStore.peers.values()).filter(
```

Comment:
probably a small thing, but `peerStore.peers` has been an expensive call in the past.

Reply:
What alternatives does libp2p offer to get a full list of available peers? If this call is expensive we can consider using an alternative peerStore

Reply:
libp2p doesn't provide anything other than a few Map objects, all keyed by peerIdStr
If you look at the implementation, you should see how you can get the keys w/o the extra bloat.

Ideally, eventually libp2p moves entirely to peerIdStr identifiers instead of PeerId objects.

Reply:
You warned https://github.com/ChainSafe/lodestar/pull/2238